### ISSUE TITLE
XL: GetObject returns bytes mismatch

### ISSUE THREAD

**Issue by harshavardhana** [minio/minio#1792]
```
Goroutine: 55
--- FAIL: TestFunctional (8.27s)
        api-functional-v4_test.go:1221: Error: bytes mismatch.
FAIL

Goroutine: 448
--- FAIL: TestFunctional (8.39s)
        api-functional-v4_test.go:1221: Error: bytes mismatch.
FAIL

Goroutine: 70
--- FAIL: TestFunctional (8.38s)
        api-functional-v4_test.go:1221: Error: bytes mismatch.
FAIL

Goroutine: 258
--- FAIL: TestFunctional (8.49s)
        api-functional-v4_test.go:1221: Error: bytes mismatch.
FAIL

Goroutine: 288
--- FAIL: TestFunctional (8.43s)
        api-functional-v4_test.go:1221: Error: bytes mismatch.
FAIL

Goroutine: 164
--- FAIL: TestFunctional (8.58s)
        api-functional-v4_test.go:1221: Error: bytes mismatch.
FAIL
```

**Comment by harshavardhana** [comment]
This bug was introduced with current storage API implementation,  still working on a fix.  It is the erasure decoding which is causing the problem.

**Comment by balamurugana** [comment]
I am able to reproduce this with current head of rewrite-xl branch.  sometimes it gives same data back.

``` bash
$ mc mb myminio/mybucket
Bucket created successfully  myminio/mybucket .

$ mc cp --no-color file4 myminio/mybucket/file
file4: 7.80 MB / 7.80 MB  100.00 % 23.40 MB0

$ mc ls -r myminio/mybucket
[2016-05-29 04:51:12 PDT] 7.8MiB file

$ mc cp myminio/mybucket/file /tmp/file
http://127.0.0.1:60000/mybucket/file: 7.80 MB / 7.80 MB  100.00 % 162.12 MB0

$ md5sum /tmp/file file4
77fd91fba74d5b14fc9ebe586dc051b4  /tmp/file
497ec12f970c135230d5ff56def9a52e  file4

$ mc cp --no-color file1 myminio/mybucket/file
file1: 18.22 KB / 18.22 KB  100.00 % 1.14 MB0

$ mc ls -r myminio/mybucket
[2016-05-29 04:51:12 PDT]  18KiB file

$ mc cp myminio/mybucket/file /tmp/file
http://127.0.0.1:60000/mybucket/file: 18.22 KB / 18.22 KB  100.00 % 994.50 KB0

$ md5sum /tmp/file file1
1002e66d619568f6fe0818374d6a760f  /tmp/file
1002e66d619568f6fe0818374d6a760f  file1

$ mc cp --no-color file3 myminio/mybucket/file
file3: 4.32 MB / 4.32 MB  100.00 % 24.11 MB0

$ mc ls -r myminio/mybucket
[2016-05-29 04:51:12 PDT] 4.3MiB file

$ mc cp myminio/mybucket/file /tmp/file
http://127.0.0.1:60000/mybucket/file: 4.32 MB / 4.32 MB  100.00 % 111.06 MB0

$ md5sum /tmp/file file3
ee73a9c97928bd76ff68fb2e07f7fea0  /tmp/file
76bf8cb11c945982c2e7efadaf1f0d76  file3

$ mc rm myminio/mybucket/file
Removed  myminio/mybucket/file .

$ mc cp --no-color file1 myminio/mybucket/file
file1: 18.22 KB / 18.22 KB  100.00 % 1.24 MB0

$ mc ls -r myminio/mybucket
[2016-05-29 04:51:12 PDT]  18KiB file

$ mc cp myminio/mybucket/file /tmp/file1
http://127.0.0.1:60000/mybucket/file: 18.22 KB / 18.22 KB  100.00 % 1.01 MB0

$ md5sum /tmp/file file1
ee73a9c97928bd76ff68fb2e07f7fea0  /tmp/file
1002e66d619568f6fe0818374d6a760f  file1

```

**Comment by harshavardhana** [comment]
> I am able to reproduce this with current head of rewrite-xl branch. sometimes it gives same data back.
> 
> $ mc mb myminio/mybucket
> Bucket created successfully  myminio/mybucket .
> 
> $ mc cp --no-color file4 myminio/mybucket/file
> file4: 7.80 MB / 7.80 MB  100.00 % 23.40 MB0
> 
> $ mc ls -r myminio/mybucket
> [2016-05-29 04:51:12 PDT] 7.8MiB file
> 
> $ mc cp myminio/mybucket/file /tmp/file
> http://127.0.0.1:60000/mybucket/file: 7.80 MB / 7.80 MB  100.00 % 162.12 MB0
> 
> $ md5sum /tmp/file file4
> 77fd91fba74d5b14fc9ebe586dc051b4  /tmp/file
> 497ec12f970c135230d5ff56def9a52e  file4
> 
> $ mc cp --no-color file1 myminio/mybucket/file
> file1: 18.22 KB / 18.22 KB  100.00 % 1.14 MB0
> 
> $ mc ls -r myminio/mybucket
> [2016-05-29 04:51:12 PDT]  18KiB file
> 
> $ mc cp myminio/mybucket/file /tmp/file
> http://127.0.0.1:60000/mybucket/file: 18.22 KB / 18.22 KB  100.00 % 994.50 KB0
> 
> $ md5sum /tmp/file file1
> 1002e66d619568f6fe0818374d6a760f  /tmp/file
> 1002e66d619568f6fe0818374d6a760f  file1
> 
> $ mc cp --no-color file3 myminio/mybucket/file
> file3: 4.32 MB / 4.32 MB  100.00 % 24.11 MB0
> 
> $ mc ls -r myminio/mybucket
> [2016-05-29 04:51:12 PDT] 4.3MiB file
> 
> $ mc cp myminio/mybucket/file /tmp/file
> http://127.0.0.1:60000/mybucket/file: 4.32 MB / 4.32 MB  100.00 % 111.06 MB0
> 
> $ md5sum /tmp/file file3
> ee73a9c97928bd76ff68fb2e07f7fea0  /tmp/file
> 76bf8cb11c945982c2e7efadaf1f0d76  file3
> 
> $ mc rm myminio/mybucket/file
> Removed  myminio/mybucket/file .
> 
> $ mc cp --no-color file1 myminio/mybucket/file
> file1: 18.22 KB / 18.22 KB  100.00 % 1.24 MB0
> 
> $ mc ls -r myminio/mybucket
> [2016-05-29 04:51:12 PDT]  18KiB file
> 
> $ mc cp myminio/mybucket/file /tmp/file1
> http://127.0.0.1:60000/mybucket/file: 18.22 KB / 18.22 KB  100.00 % 1.01 MB0
> 
> $ md5sum /tmp/file file1
> ee73a9c97928bd76ff68fb2e07f7fea0  /tmp/file
> 1002e66d619568f6fe0818374d6a760f  file1

There is a possible fix #1793 - there is some multipart issue still - can you review the patch?
